subsys/nfc/tnep/tag.c

```diff
@@ -0,0 +1,700 @@
+
+	return err;
+}
+
+static int tnep_tx_msg_add_rec(const struct nfc_ndef_record_desc *record)
+{
+	int err = 0;
+
+	if (!record) {
+		return -EINVAL;
+	}
+
+	err = nfc_ndef_msg_record_add(tnep.msg,
+				      record);
+
+	return err;
+}
+
+static int tnep_tx_msg_add_rec_status(enum nfc_tnep_status_value status)
+{
+	int err = 0;
+
+	if (status == NFC_TNEP_STATUS_PROTOCOL_ERROR) {
+		tnep_tx_msg_clear();
+	}
+
+	status_record.status = status;
+
+	err = tnep_tx_msg_add_rec(&NFC_NDEF_TNEP_RECORD_DESC(status_record));
+
+	return err;
+}
+
+static int tnep_tx_initial_msg_set(void)
+{
+	int err = 0;
+
+	tnep_tx_msg_clear();
+
+	for (size_t i = 0; i < tnep.svc_cnt; i++) {
+		err = tnep_tx_msg_add_rec(tnep.svc[i].ndef_record);
+		if (err) {
+			return err;
+		}
+	}
+
+	if (!tnep.records) {
+		return tnep_tx_msg_encode();
+	}
+
+	for (size_t i = 0; i < tnep.records_cnt; i++) {
+		err = tnep_tx_msg_add_rec(&tnep.records[i]);
+		if (err) {
+			return err;
+		}
+	}
+
+	return tnep_tx_msg_encode();
+}
+
+static bool ndef_check_rec_type(const struct nfc_ndef_record_desc *record,
+				const u8_t *type_field,
+				const u32_t type_field_length)
+{
+	int cmp_res;
+
+	if (record->type_length != type_field_length) {
+		return false;
+	}
+
+	cmp_res = memcmp(record->type, type_field, type_field_length);
+
+	return (cmp_res == 0);
+}
+
+static int tnep_svc_set_active(const u8_t *uri_name, size_t uri_length)
+{
+
+	for (size_t i = 0; i < tnep.svc_cnt; i++) {
+		if ((tnep.svc[i].parameters->uri_length == uri_length) &&
+		    !memcmp(uri_name,
+			    tnep.svc[i].parameters->uri, uri_length)) {
+			tnep.svc_active = &tnep.svc[i];
+
+			return TNEP_SVC_SELECTED;
+		}
+	}
+
+	tnep.svc_active = NULL;
+
+	return TNEP_SVC_DESELECTED;
+}
+
+static int tnep_svc_select_from_msg(void)
+{
+	int err;
+	enum tnep_svc_record_status svc_status = TNEP_SVC_NOT_FOUND;
+	const struct nfc_ndef_msg_desc *msg_p;
+	u8_t desc_buf[NFC_NDEF_PARSER_REQIRED_MEMO_SIZE_CALC(
+				CONFIG_NFC_TNEP_RX_MAX_RECORD_CNT)];
+	size_t desc_buf_len = sizeof(desc_buf);
+
+	err = nfc_ndef_msg_parse(desc_buf, &desc_buf_len, tnep.rx.data,
+				 &tnep.rx.len);
+	if (err) {
+		return -EINVAL;
+	}
+
+	msg_p = (struct nfc_ndef_msg_desc *) desc_buf;
+
+	/* Search for service records */
+	for (size_t i = 0; i < msg_p->record_count; i++) {
+		bool select_rec = ndef_check_rec_type(msg_p->record[i],
+						      nfc_ndef_tnep_rec_type_svc_select,
+						      NFC_NDEF_TNEP_REC_TYPE_LEN);
+
+		if (select_rec) {
+			/* Set service from record as active (or deselect) */
+			u8_t bin_rec[CONFIG_NFC_TNEP_RX_MAX_RECORD_SIZE];
+			u32_t bin_rec_len = sizeof(bin_rec);
+			struct nfc_ndef_tnep_rec_svc_select service;
+
+			memset(bin_rec, 0x00, bin_rec_len);
+
+			msg_p->record[i]->payload_constructor(
+					msg_p->record[i]->payload_descriptor,
+					bin_rec, &bin_rec_len);
+
+			service.uri_len = bin_rec[0];
+			service.uri = &bin_rec[1];
+
+			svc_status = tnep_svc_set_active(service.uri,
+							 service.uri_len);
+		}
```

Comment:
Why don't we break here? There can only be 1 Service Select record in a Service Select message.

Reply:
Added. I do not see any reason why it was not there.